### PR TITLE
Define the iota combinator for arrays

### PR DESCRIPTION
**Pull request by DanielRrr**
The list type has a convenient combinator `iota`, whereas the lean standard library lacks the iota counterpart for arrays. In this patch, I introduce it. We may use this module further for more array related functions

I also introduce the list-style monad instance for `Array`, which is needed for some parts of Nova as well.

### CODE REVIEW

File: YatimaStdLib/Array.lean [hunk 1]
```diff
@@ -0,0 +1,13 @@
+-- generates the array of nats from 0,...,n by a given n
+def Array.iota (n : Nat) : Array Nat :=
+  match n with
+    | 0 => #[0]
+    | k+1 => Array.push (iota k) (k + 1)
```
Comment (arthurpaulino) [inline YatimaStdLib/Array.lean:5]:
```suggestion
  Array.mk (List.range n) |>.push n
```
Let's make use of `List.range`, which is tail recursive

**Review by arthurpaulino — CHANGES_REQUESTED** [review]

File: YatimaStdLib/Array.lean [hunk 2]
```diff
@@ -0,0 +1,11 @@
+-- generates the array of nats from 0,...,n by a given n
```
Comment (arthurpaulino) [inline YatimaStdLib/Array.lean:1]:
```suggestion
/-- Generates the array of nats from 0,...,n by a given n -/
```

File: YatimaStdLib/Array.lean [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+def Array.iota (n : Nat) : Array Nat :=
+  Array.mk (List.range n) |>.push n
+
+def Array.join (l : Array (Array A)) : Array A :=
+  Array.foldr (. ++ .) #[] l
+
+instance : Monad Array where
+  map := Array.map
+  pure x := #[x]
+  bind l f := Array.join $ Array.map f l
```
Comment (arthurpaulino) [inline YatimaStdLib/Array.lean:11]:
```suggestion
  bind l f := Array.join $ Array.map f l

```

**Review by arthurpaulino — APPROVED** [review]

File: YatimaStdLib/Array.lean (hunk 3 above)
Comment (DanielRrr) [reply YatimaStdLib/Array.lean:11]: I don't get the difference

Comment (arthurpaulino) [reply YatimaStdLib/Array.lean:11]: Just adding an empty line at the end of the file for Git friendliness

**Pull request merged**